ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/FollowerPromptness.hs

```diff
@@ -0,0 +1,214 @@
+-- containing it.
+module Test.Ouroboros.Storage.ChainDB.FollowerPromptness (tests) where
+
+import           Control.Monad (forever)
+import           Control.Monad.IOSim (runSimOrThrow)
+import           Control.Tracer (Tracer (..), contramapM, traceWith)
+import           Data.Foldable (for_)
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Set (Set)
+import qualified Data.Set as Set
+
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+import qualified Ouroboros.Network.MockChain.Chain as Chain
+
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.Storage.ChainDB.API (ChainDB)
+import qualified Ouroboros.Consensus.Storage.ChainDB.API as ChainDB
+import qualified Ouroboros.Consensus.Storage.ChainDB.API.Types.InvalidBlockPunishment as Punishment
+import           Ouroboros.Consensus.Storage.ChainDB.Impl (ChainDbArgs (..))
+import qualified Ouroboros.Consensus.Storage.ChainDB.Impl as ChainDBImpl
+import           Ouroboros.Consensus.Util.Condense (Condense (..))
+import           Ouroboros.Consensus.Util.Enclose
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Test.Util.ChainDB
+import           Test.Util.ChainUpdates
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock
+import           Test.Util.Tracer (recordingTracerTVar)
+
+tests :: TestTree
+tests = testGroup "FollowerPromptness"
+    [ testProperty "followerPromptness" prop_followerPromptness
+    ]
+
+prop_followerPromptness :: FollowerPromptnessTestSetup -> Property
+prop_followerPromptness fpts =
+    counterexample ("Trace:\n" <> unlines (ppTrace <$> traceByTime)) $
+    counterexample (condense fpts) $
+    counterexample ("Instruction timings: " <> condense followerInstrTimings) $
+    counterexample ("Failed to pipeline: " <> condense notPipelined)
+    (null notPipelined) .&&.
+    counterexample ("Not processed: " <> condense unprocessed)
+    (null unprocessed)
+  where
+    FollowerPromptnessOutcome{..} =
+      runSimOrThrow $ runFollowerPromptnessTest fpts
+
+    -- Hashes of tentative headers which were not immediately emitted as a
+    -- follower instruction.
+    notPipelined =
+        tentativeHeadersWithoutFollowUp
+          0
+          followerInstrTimings
+
+    -- Hashes of tentative header which were not processed (i.e. made obsolete
+    -- due to adoption, or identified as a trap).
+    unprocessed =
+        tentativeHeadersWithoutFollowUp
+          artificialDelay
+          tentativeHeaderUnsetTimings
+
+    -- Given a collection of timestamped hashes (considered as follow-up events
+    -- of a specific hash), return the timestamped tentative header hashes which
+    -- are not contained therein after the given delay.
```

Comment:
Maybe add a sentence explicitly emphasizing that this is in `io-sim` with carefully controlled passage of time and so we know _exactly_ what timestamp the expected events will have.

When I first read this comment, I wondered if "after the given delay" meant "within the given delay", but it's not, your wording was precise. It's just an unusual time semantics to come across in general (but understandable in the `io-sim` context).

Reply:
I added this part to the module header:
```haskell
-- This test is making use of io-sim to measure and check the *exact* timings of
-- various events. In particular, we can really rely on something occuring at a
-- specific point in time, compared to just a plausible range as would be
-- necessary with ordinary wall-clock time.
```